### ISSUE TITLE
Existing files from 0.98 are not upgraded automatically by 0.101

### ISSUE THREAD

**Issue by unusuallyreticent** [realm/realm-cocoa#3581]
## Goals
Upgrade from 0.98 to .102

What do you want to achieve?
Low stress.

## Expected Results
No stress

## Actual Results
Stress

What did happened instead?  

If the realm Configuration deleteRealmIfMigrationNeeded is false, the io.realm domain error code: 2 is returned in a catch with ‘Unable to open a realm at path …’
If deleteRealmIfMigrationNeeded is true, the application will crash with a RealmFileException ‘Unable to open a realm at path …’

The Realm Browser App states the file is an older version and offers to upgrade it. It upgrades fine.

Seems to be a variant  of report 3577, or maybe I am interpreting the following change notes incorrectly:

0.101.0 Release notes (2016-05-04) 
Files written by this version of Realm cannot be read by older versions of Realm. **Existing files will automatically be upgraded when they are opened.**

## Steps to Reproduce
Create a realm file with 0.98. Open it with a readonly configuration in 0.102.

## Version of Realm and Tooling
Realm version: 0.102
Xcode version: 7.3
iOS/OSX version: 9.3

**Comment by bdash** [comment]
I think this is a case of the release notes not being as clear as they should be. The [release blog post for v0.101](https://realm.io/news/realm-objc-swift-0.101.0/) does a better job:

> This new property type format required an update to the overall Realm file format, meaning that Realm files read or written by this version cannot be opened with previous versions of Realm. Existing files will automatically be upgraded when they are opened, except for read-only files bundled with your app, which you can upgrade by opening them in the Realm Browser. Please be careful about shipping updates to your existing apps! 🚨

Files cannot be updated automatically when opened as read-only since updating to the new format requires modifying the file.

**Comment by unusuallyreticent** [comment]
Oh!
The realm file is not actually read-only - as a file and it is not bundled with the App. Rather, it is created by the App in a sub-folder of the Application Support folder from downloaded XML data files. The read-only part is the readonly parameter passed in the Realm Configuration.

**Comment by bdash** [comment]
Understood. We interpret the read-only setting on the configuration as indicating that we should not modify the file, so we don't attempt to update it to the latest format. The blog post could be clearer on this as it does sound like there's something specific about Realm files bundled with your app. The way to deal with the upgrade is also different if it's not a bundled file (you'll need to either remove and recreate the files, or open them read-write to allow the file format to be upgraded).

**Comment by unusuallyreticent** [comment]
Understood. I did not consider the readOnly parameter would bar a realm file from being auto-upgraded if a structure version change was detected. I can't exactly see the logic though. On iOS, for as many reasons the file should not be upgraded, there are more reasons why it should be upgraded. It's not a schema change.

**Comment by unusuallyreticent** [comment]
Out of curiosity, within an iOS app deliberately upgraded with realm 0.101upwards, what reason could there be to NOT auto-upgrade a realm file irrespective of the readOnly mode in the Configuration?

**Comment by bdash** [comment]
Realm currently only supports opening files in two modes: read-write, and read-only. In read-only mode the file is never modified in any way. Having a mode that is called "read-only" but still modified the file would be rather misleading. It sounds like what you're after is a third mode where the file would be writeable only for system transactions such as file format upgrades, but read-only from the application's point of view. I can see the value in that, but I don't think it'd desirable to repurpose the term "read-only" for it.

**Comment by unusuallyreticent** [comment]
Exactly. I think of readOnly as 'my' intent (don't want to modify the object graph) and you think of readOnly as don't mess with the on disk file. However, whatever my intent is, the older file is useless if the realm engine is updated. IMO, on disk structure upgrades need to be enabled/disabled separately to the requested opening mode. I've used an autoreleasepool in a little loop to first try opening readOnly and on failure do a little dance through readWrite and readOnly again. If that fails, data is downloaded again and the realm files are deleted and rebuilt from scratch. 
The App is used in cockpits. Sometimes, Airline IT Managers update the App but don't run it. It's not so good if existing data files don't open :-(
